app/javascript/mastodon/store/typed_functions.ts

```diff
@@ -13,8 +15,190 @@ export interface AsyncThunkRejectValue {
+type DiscardApiResult = typeof discardApiResultInPayload;
+
+type OnData<ApiResult, Return> = (
+  data: ApiResult,
+  api: AppThunkApi & {
+    discardApiResults: DiscardApiResult;
+  },
+) => Return | DiscardApiResult | Promise<Return | DiscardApiResult>;
+
+// Overload when there is no `onData` method, the payload is the API result
+export function createDataLoadingThunk<
+  ApiResult,
+  Args extends readonly unknown[],
+>(
+  name: string,
+  loadData: (...args: Args) => Promise<ApiResult>,
+  thunkOptions?: AppThunkOptions,
+): ReturnType<typeof createThunk<Args, ApiResult>>;
+
+// Overload when the `onData` method returns discardApiResultInPayload, then the payload is empty
+export function createDataLoadingThunk<
+  ApiResult,
+  Args extends readonly unknown[],
+>(
+  name: string,
+  loadData: (...args: Args) => Promise<ApiResult>,
+  onDataOrThunkOptions?: AppThunkOptions | OnData<ApiResult, DiscardApiResult>,
+  thunkOptions?: AppThunkOptions,
+): ReturnType<typeof createThunk<Args, void>>;
+
+// Overload when the `onData` method returns discardApiResultInPayload, then the mayload is the API result
+export function createDataLoadingThunk<
+  ApiResult,
+  Args extends readonly unknown[],
+>(
+  name: string,
+  loadData: (...args: Args) => Promise<ApiResult>,
+  onDataOrThunkOptions?: AppThunkOptions | OnData<ApiResult, void>,
+  thunkOptions?: AppThunkOptions,
+): ReturnType<typeof createThunk<Args, ApiResult>>;
+
+// Overload when there is an `onData` method returning something
+export function createDataLoadingThunk<
+  ApiResult,
+  Args extends readonly unknown[],
+  Returned,
+>(
+  name: string,
+  loadData: (...args: Args) => Promise<ApiResult>,
+  onDataOrThunkOptions?: AppThunkOptions | OnData<ApiResult, Returned>,
+  thunkOptions?: AppThunkOptions,
+): ReturnType<typeof createThunk<Args, Returned>>;
+
+/**
+ * This function creates a Redux Thunk that handles loading data asynchronously (usually from the API), dispatching `pending`, `fullfilled` and `rejected` actions.
+ *
+ * You can run a callback on the API results to either dispatch side effects or modify the payload.
+ *
+ * It is a wrapper around RTK's [`createAsyncThunk`](https://redux-toolkit.js.org/api/createAsyncThunk)
+ * @param name Prefix for the actions types
+ * @param loadData Function that loads the data. It's arguments will become the thunk's arguments
+ * @param onDataOrThunkOptions
+ *   Callback called on the results from `loadData`.
+ *
+ *   First argument will be the return from `loadData`.
+ *
+ *   Second argument is an object with: `dispatch`, `getState` and `discardApiResults`.
+ *   It can return:
+ *   - `undefined` (or no explicit return), meaning that the API results will be the payload
+ *   - `discardApiResults` to discard the API results and return an empty payload
+ *   - anything else, which will be the payload
+ *
+ *   You can also omit this parameter and pass `thunkOptions` directly
+ * @param thunkOptionsParams
```

Comment:
```suggestion
 * @param thunkOptions
```
To match the usage of thunkOptions in the line above